pallets/automation-time/src/lib.rs

```diff
@@ -1343,27 +1325,74 @@ pub mod pallet {
 			Ok(())
 		}
 
-		/// Reschedules an existing task for a given number of execution times
-		fn reschedule_existing_task(
+		fn reschedule_or_remove_task(
 			task_id: TaskId<T>,
-			task: &TaskOf<T>,
-			execution_times: Vec<UnixTime>,
-		) -> Result<(), DispatchError> {
-			let new_executions = execution_times.len().try_into().unwrap();
+			task: &mut TaskOf<T>,
+			dispatch_error: Option<DispatchError>,
+		) {
+			let remove_on_err = |task: &TaskOf<T>, err: DispatchError| {
+				Self::deposit_event(Event::<T>::TaskNotRescheduled {
+					who: task.owner_id.clone(),
+					task_id,
+					error: err,
+				});
+				AccountTasks::<T>::remove(task.owner_id.clone(), task_id);
+			};
 
-			let fee = Self::calculate_execution_fee(&task.action, new_executions)?;
-			T::FeeHandler::can_pay_fee(&task.owner_id, fee.clone())
-				.map_err(|_| Error::<T>::InsufficientBalance)?;
+			if let Some(e) = dispatch_error {
+				remove_on_err(task, e);
+			} else {
+				let owner_id = task.owner_id.clone();
+				match Self::reschedule_existing_task(task_id, task) {
+					Ok(_) => {
+						Self::deposit_event(Event::<T>::TaskRescheduled { who: owner_id, task_id });
+					},
+					Err(e) => {
+						remove_on_err(task, e);
+					},
+				};
+			}
+		}
 
-			Self::insert_scheduled_tasks(task_id, task, execution_times.clone())?;
+		fn reschedule_existing_task(task_id: TaskId<T>, task: &mut TaskOf<T>) -> DispatchResult {
+			match task.schedule {
```

Comment:
I think you can use an `if let` here

Reply:
I opted for a `match` instead of an `if let` because I wanted to demonstrate that this code was exhaustive on all permutations of `Schedule` in the future if we add a new variant to `Schedule` the compiler will flag that the new variant should specify if/how it should be rescheduled. We'll lose that nice compiler feature if this is converted to an `if let`.